src/dvc_data/hashfile/db/migrate.py

```diff
@@ -0,0 +1,85 @@
+from functools import partial, wraps
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, NamedTuple, Tuple
+
+from dvc_objects.executors import ThreadPoolExecutor
+from dvc_objects.fs.callbacks import DEFAULT_CALLBACK
+
+if TYPE_CHECKING:
+    from dvc_objects.fs.base import FileSystem
+    from dvc_objects.fs.callbacks import Callback
+
+    from . import HashFileDB
+
+
+class PreparedMigration(NamedTuple):
+    src: "HashFileDB"
+    dest: "HashFileDB"
+    paths: List[str]
+    oids: List[str]
+
+
+def migrate(
+    migration: "PreparedMigration", callback: "Callback" = DEFAULT_CALLBACK
+) -> int:
+    """Migrate objects from one HashFileDB to another.
+
+    Files from src will be re-hashed and transferred to dest with hardlinking
+    enabled.
+    """
+    src, dest, paths, oids = migration
+    return dest.add(paths, src.fs, oids, hardlink=True, callback=callback)
+
+
+def prepare(
```

Comment:
I separated `prepare` and `migrate` because from the DVC UI perspective it's two distinct operations with separate progress bars (one for the hashing operation and one for the transfer)

Reply:
Looks pretty nice. Need to do the same in checkout and fetch later 🙂